File: trunk/managed/Warp3D/Warp3DVB/warp_RenderPipeline.vb
' Created by: X
' http://www.createdbyx.com/
' ----------------------
' C# to VB.NET conversion of this code was done by hand and with the assistance of the 
' Convert C# to VB .NET application availible at http://www.kamalpatel.net/ConvertCSharp2VB.aspx
' ConvertCSharp2VB utility was developed by: Kamal Patel (http://www.KamalPatel.net) 
' ----------------------
' Thanks to Alan Simes ( http://alansimes.blogdns.net/ ) for creating the Warp3D library.
' http://alansimes.blogdns.net/cs/files/default.aspx
' ----------------------


Imports System
Imports System.Collections

Namespace Rednettle.Warp3D
    Public Class warp_RenderPipeline
        Implements IDisposable
        Public screen As warp_Screen
        Dim scene As warp_Scene
        Public lightmap As warp_Lightmap

        Private resizingRequested As Boolean = False
        Private requestedWidth As Integer
        Private requestedHeight As Integer
        Private splitOffset As Double
        Public useId As Boolean = False

        Dim rasterizer As warp_Rasterizer

        Dim opaqueQueue As ArrayList = New ArrayList()
        Dim transparentQueue As ArrayList = New ArrayList()

        Dim zFar As Integer = &HFFFFFF

        Public zBuffer() As Integer
        Public idBuffer() As Integer


        Public Sub New(ByVal scene As warp_Scene, ByVal w As Integer, ByVal h As Integer)
            Me.scene = scene

            screen = New warp_Screen(w, h)
            ReDim zBuffer(screen.width * screen.height)
            rasterizer = New warp_Rasterizer(Me)
        End Sub

        Public Function getFPS() As Single
            Return (screen.FPS * 100) / 100
        End Function

        Public Sub buildLightMap()
			if lightmap is nothing
                lightmap = New warp_Lightmap(scene)
            Else
                lightmap.rebuildLightmap()
            End If

            rasterizer.loadLightmap(lightmap)
        End Sub

        Public Sub useIdBuffer(ByVal useId As Boolean)
            Me.useId = useId
            If useId Then
                ReDim idBuffer(screen.width * screen.height)
            Else
                idBuffer = Nothing
            End If
        End Sub

        Public Sub render(ByVal cam As warp_Camera)
            rasterizer.rebuildReferences(Me)

            warp_Math.clearBuffer(zBuffer, zFar)
            '//System.Array.Copy(screen.zBuffer,0,zBuffer,0,zBuffer.Length);

            If useId Then
                warp_Math.clearBuffer(idBuffer, -1)
            End If
            If scene.environment.background IsNot Nothing Then
                screen.drawBackground(scene.environment.background, 0, 0, screen.width, screen.height)
            Else
                screen.clear(scene.environment.bgcolor)
            End If

            cam.setScreensize(screen.width, screen.height)
            scene.prepareForRendering()
            emptyQueues()

            '// Project
            Dim m As warp_Matrix = warp_Matrix.multiply(cam.getMatrix(), scene.matrix)
            Dim nm As warp_Matrix = warp_Matrix.multiply(cam.getNormalMatrix(), scene.normalmatrix)
            Dim vertexProjection, normalProjection As warp_Matrix
            Dim obj As warp_Object
            Dim t As warp_Triangle
            Dim v As warp_Vertex
            Dim w As Integer = screen.width
            Dim h As Integer = screen.height
            For id As Integer = scene.objects - 1 To 0 Step -1
                obj = scene.wobject(id)
                If obj.visible Then
                    vertexProjection = obj.matrix.getClone()
                    normalProjection = obj.normalmatrix.getClone()
                    vertexProjection.transform(m)
                    normalProjection.transform(nm)

                    For i As Integer = obj.vertices - 1 To 0 Step -1
                        v = obj.fastvertex(i)
                        v.project(vertexProjection, normalProjection, cam)
                        v.clipFrustrum(w, h)
                    Next
                    For i As Integer = obj.triangles - 1 To 0 Step -1
                        t = obj.fasttriangle(i)
                        t.project(normalProjection)
                        t.clipFrustrum(w, h)
                        enqueueTriangle(t)
                    Next
                End If
                    Next

            '//screen.lockImage();

            Dim tri() As warp_Triangle
            tri = getOpaqueQueue()
            If tri IsNot Nothing Then
                For i As Integer = tri.GetLength(0) - 1 To 0 Step -1
                    rasterizer.loadMaterial(tri(i).parent.material)
                    rasterizer.render(tri(i))
                Next
            End If

            tri = getTransparentQueue()
            If tri IsNot Nothing Then
                For i As Integer = 0 To tri.GetLength(0) - 1
                    rasterizer.loadMaterial(tri(i).parent.material)
                    rasterizer.render(tri(i))
                Next
            End If

            '//screen.unlockImage();
        End Sub


        Private Sub performResizing()
            resizingRequested = False
            'screen.resize(requestedWidth, requestedHeight);

            zBuffer = New Integer(screen.width * screen.height) {}
            If useId Then
                idBuffer = New Integer(screen.width * screen.height) {}
            End If
        End Sub

        ' Triangle sorting
        Private Sub emptyQueues()
            opaqueQueue.Clear()
            transparentQueue.Clear()
        End Sub


        Private Sub enqueueTriangle(ByVal tri As warp_Triangle)
            If tri.parent.material Is Nothing Then
                Return
            End If
            If tri.visible = False Then
                Return
            End If
            If (tri.parent.material.transparency = 255) AndAlso (tri.parent.material.reflectivity = 0) Then
                Return
            End If

            If tri.parent.material.transparency > 0 Then
                transparentQueue.Add(tri)
            Else
                opaqueQueue.Add(tri)
            End If
        End Sub

        Private Function getOpaqueQueue() As warp_Triangle()
            If opaqueQueue.Count = 0 Then Return Nothing

            Dim enumerator As IEnumerator = opaqueQueue.GetEnumerator()
            Dim tri(opaqueQueue.Count) As warp_Triangle

            Dim id As Integer = 0
            While enumerator.MoveNext()
                tri(id) = CType(enumerator.Current, Warp3D.warp_Triangle)
                id += 1
            End While

            Return sortTriangles(tri, 0, tri.GetLength(0) - 1)
        End Function

        Private Function getTransparentQueue() As warp_Triangle()
            If transparentQueue.Count = 0 Then Return Nothing
            Dim enumerator As IEnumerator = transparentQueue.GetEnumerator()
            Dim tri(transparentQueue.Count) As warp_Triangle

            Dim id As Integer = 0
            While enumerator.MoveNext()
                tri(id) = CType(enumerator.Current, Warp3D.warp_Triangle)
                id += 1
            End While

            Return sortTriangles(tri, 0, tri.GetLength(0) - 1)
        End Function

        Private Function sortTriangles(ByVal tri() As warp_Triangle, ByVal L As Integer, ByVal R As Integer) As warp_Triangle()
            Dim m As Single = (tri(L).dist + tri(R).dist) / 2
            Dim i As Integer = L
            Dim j As Integer = R
            Dim temp As warp_Triangle

            Do
                While tri(i).dist > m
                    i += 1
                End While
                While tri(j).dist < m
                    j -= 1
                End While

                If i <= j Then
                    temp = tri(i)
                    tri(i) = tri(j)
                    tri(j) = temp
                    i += 1
                    j -= 1
                End If
                
            Loop While j >= i

            If L < j Then sortTriangles(tri, L, j)
            If R > i Then sortTriangles(tri, i, R)

            Return tri
        End Function

        Public Sub Dispose() Implements IDisposable.Dispose

        End Sub
    End Class
End Namespace

File: trunk/managed/Warp3Dmod/warp_RenderPipeline.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace Warp3D
{
    /// <summary>
    /// Summary description for warp_RenderPipeline.
    /// </summary>
    public class warp_RenderPipeline: IDisposable
    {
        public warp_Screen screen;
        warp_Scene scene;
        public warp_Lightmap lightmap;

        public bool useId = false;

        warp_Rasterizer rasterizer;

        ArrayList transparentQueue = new ArrayList(16384);

        int zFar = int.MaxValue;

        public int[] zBuffer;

        public warp_RenderPipeline(warp_Scene scene, int w, int h)
        {
            this.scene = scene;

            screen = new warp_Screen(w, h);
            zBuffer = new int[screen.width * screen.height];
            rasterizer = new warp_Rasterizer(this);
        }

        public void buildLightMap()
        {
            if(lightmap == null)
            {
                lightmap = new warp_Lightmap(scene);
            }
            else
            {
                lightmap.rebuildLightmap();
            }

            rasterizer.loadLightmap(lightmap);
        }

        public void render(warp_Camera cam)
        {
            rasterizer.rebuildReferences(this);

            warp_Math.clearBuffer(zBuffer, zFar);
            //System.Array.Copy(screen.zBuffer,0,zBuffer,0,zBuffer.Length);

            if(scene.environment.background != null)
            {
                screen.drawBackground(scene.environment.background, 0, 0, screen.width, screen.height);
            }
            else
            {
                screen.clear(scene.environment.bgcolor);
            }

            cam.setScreensize(screen.width, screen.height);
            scene.prepareForRendering();
            emptyQueues();

            // Project
            warp_Matrix m = warp_Matrix.multiply(cam.getMatrix(), scene.matrix);
            warp_Matrix nm = warp_Matrix.multiply(cam.getNormalMatrix(), scene.normalmatrix);
            warp_Matrix vertexProjection, normalProjection;
            warp_Object obj;
            warp_Triangle t;
            warp_Vertex v;
            warp_Material objmaterial;
            const double log2inv = 1.4426950408889634073599246810019;
            int w = screen.width;
            int h = screen.height;
            int minx;
            int miny;
            int maxx;
            int maxy;

            for(int id = 0 ; id < scene.objects; ++id)
            {
                obj = scene.wobject[id];
                objmaterial = obj.material;
                if(objmaterial == null)
                    continue;
                if(!obj.visible)
                    continue;
                if(objmaterial.opaque && objmaterial.reflectivity == 0)
                    continue;

                vertexProjection = obj.matrix.getClone();
                normalProjection = obj.normalmatrix.getClone();
                vertexProjection.transform(m);
                normalProjection.transform(nm);
                minx = int.MaxValue;
                miny = int.MaxValue;
                maxx = int.MinValue;
                maxy = int.MinValue;

                for(int i = 0; i < obj.vertices; ++i)
                {
                    v = obj.fastvertex[i];
                    v.project(vertexProjection, normalProjection, cam);
                    v.clipFrustrum(w, h);
                    if(minx > v.x)
                        minx = v.x;
                    if(maxx < v.x)
                        maxx = v.x;
                    if(miny > v.y)
                        miny = v.y;
                    if(maxy < v.y)
                        maxy = v.y;
                }
                maxx -= minx;
                maxy -= miny;
                if(maxy > maxx)
                    maxx = maxy + 1;
                else
                    maxx++;

                obj.projectedmaxMips = (int)Math.Ceiling((Math.Log(maxx) * log2inv));
                obj.cacheMaterialData();
                rasterizer.loadFastMaterial(obj);

                for(int i = 0; i < obj.triangles; ++i)
                {
                    t = obj.fasttriangle[i];
                    t.project(normalProjection);
                    t.clipFrustrum(w, h);
                    if(!t.visible)
                        continue;
                    if(objmaterial.opaque)
                    {
                        rasterizer.render(t);
                        continue;
                    }
                    transparentQueue.Add(t);
                }
            }

            //screen.lockImage();

            warp_Triangle[] tri;
            obj = null;
            tri = getTransparentQueue();
            transparentQueue.Clear();
            if(tri != null)
            {
                for(int i = 0; i < tri.GetLength(0); i++)
                {
                    if(obj != tri[i].parent)
                    {
                        obj = tri[i].parent;
                        rasterizer.loadFastMaterial(obj);
                    }
                    rasterizer.render(tri[i]);
                }
            }

            //screen.unlockImage();
        }

        private void performResizing()
        {
            //screen.resize(requestedWidth, requestedHeight);
            zBuffer = new int[screen.width * screen.height];
        }

        // Triangle sorting
        private void emptyQueues()
        {
            transparentQueue.Clear();
        }

        private warp_Triangle[] getTransparentQueue()
        {
            if(transparentQueue.Count == 0)
                return null;
            IEnumerator enumerator = transparentQueue.GetEnumerator();
            warp_Triangle[] tri = new warp_Triangle[transparentQueue.Count];

            int id = 0;
            while(enumerator.MoveNext())
                tri[id++] = (warp_Triangle)enumerator.Current;

            return sortTriangles(tri, 0, tri.GetLength(0) - 1);
        }

        private warp_Triangle[] sortTriangles(warp_Triangle[] tri, int L, int R)
        {
            //FIX: Added Index bounds checking. (Was causing random exceptions) - Created by: X
            if(L < 0)
                L = 0;
            if(L > tri.GetLength(0))
                L = tri.GetLength(0);
            if(R < 0)
                R = 0;
            if(R > tri.GetLength(0))
                R = tri.GetLength(0);
            // - Created by: X

            float m = (tri[L].distZ + tri[R].distZ) / 2;
            int i = L;
            int j = R;
            warp_Triangle temp;

            do
            {
                while(tri[i].distZ > m)
                    i++;
                while(tri[j].distZ < m)
                    j--;

                if(i <= j)
                {
                    temp = tri[i];
                    tri[i] = tri[j];
                    tri[j] = temp;
                    i++;
                    j--;
                }
            }
            while(j >= i);

            if(L < j)
                sortTriangles(tri, L, j);
            if(R > i)
                sortTriangles(tri, i, R);

            return tri;
        }

        public void Dispose()
        {
            if(screen != null)
                screen.Dispose();
            screen = null;
            zBuffer = null;
            rasterizer.clean();
            rasterizer = null;
            transparentQueue.Clear();
            transparentQueue = null;
        }
    }
}
